src/lib/storage/index/partial_hash/partial_hash_index_impl.cpp

```diff
@@ -0,0 +1,165 @@
+  }
+
+  if (_null_positions.contains(DataType{})) {
+    auto& nulls = _null_positions[DataType{}];
+    nulls.erase(std::remove_if(nulls.begin(), nulls.end(), is_to_unindex), nulls.end());
+  }
+
+  return size_before - _indexed_chunk_ids.size();
+}
+
+template <typename DataType>
+bool PartialHashIndexImpl<DataType>::indexed_null_values() const {
+  return null_cbegin() != null_cend();
+}
+
+template <typename DataType>
+BasePartialHashIndexImpl::IteratorRange PartialHashIndexImpl<DataType>::range_equals(
+    const AllTypeVariant& value) const {
+  const auto begin = _positions.find(boost::get<DataType>(value));
+  if (begin == _positions.end()) {
+    const auto end_iter = cend();
+    return std::make_pair(end_iter, end_iter);
+  }
+  auto end = begin;
+  return std::make_pair(_create_iterator(begin), _create_iterator(++end));
+}
+
+template <typename DataType>
+BasePartialHashIndexImpl::IteratorRangePair PartialHashIndexImpl<DataType>::range_not_equals(
+    const AllTypeVariant& value) const {
+  const auto range_eq = range_equals(value);
+  return std::make_pair(std::make_pair(cbegin(), range_eq.first), std::make_pair(range_eq.second, cend()));
+}
+
+template <typename DataType>
+BasePartialHashIndexImpl::Iterator PartialHashIndexImpl<DataType>::cbegin() const {
+  return _create_iterator(_positions.cbegin());
+}
+
+template <typename DataType>
+BasePartialHashIndexImpl::Iterator PartialHashIndexImpl<DataType>::cend() const {
+  return _create_iterator(_positions.cend());
+}
+
+template <typename DataType>
+BasePartialHashIndexImpl::Iterator PartialHashIndexImpl<DataType>::null_cbegin() const {
+  return _create_iterator(_null_positions.cbegin());
+}
+
+template <typename DataType>
+BasePartialHashIndexImpl::Iterator PartialHashIndexImpl<DataType>::null_cend() const {
+  return _create_iterator(_null_positions.cend());
+}
+
+template <typename DataType>
+size_t PartialHashIndexImpl<DataType>::estimate_memory_usage() const {
+  auto bytes = size_t{0u};
+
+  bytes += sizeof(_indexed_chunk_ids);  // NOLINT - Linter complains size() should be used, which would be wrong.
+  bytes += sizeof(ChunkID) * _indexed_chunk_ids.size();
+
+  bytes += sizeof(_positions);
+  bytes += sizeof(DataType) * _positions.size();
+
+  bytes += sizeof(std::vector<RowID>) * _positions.size();
+  bytes += sizeof(RowID) * std::distance(cbegin(), cend());
+
+  bytes += sizeof(_null_positions);  // NOLINT
+
+  if (_null_positions.contains(DataType{})) {
+    bytes += sizeof(std::vector<RowID>);
+    bytes += sizeof(RowID) * std::distance(null_cbegin(), null_cend());
+  }
+
+  return bytes;
+}
+
+template <typename DataType>
+std::unordered_set<ChunkID> PartialHashIndexImpl<DataType>::get_indexed_chunk_ids() const {
+  return _indexed_chunk_ids;
+}
+
+template <typename DataType>
+BasePartialHashIndexImpl::Iterator PartialHashIndexImpl<DataType>::_create_iterator(const MapIterator it) const {
+  return Iterator(std::make_unique<FlatMapIteratorImpl<DataType>>(it));
```

Comment:
Don't know if that's prettier, but can the `FatMapIterator` can have a static method that is called here:
```c++

template <typename DataType>
static FlatMapiterator from_map_iterator(const tsl::sparse_map<DataType, std::vector<RowID>>::const_iterator& it) {
  return FlatMapIterator(std::make_unique<FlatMapIteratorImpl<DataType>>(it));
}
```

Then the Impl would be completely capsuled inside the FlatMapIterator (maybe even with protected/private constructor):
```c++
return Iterator::from_map_iterator(it));
```

Reply:
We tried doing it that way, but we were not happy with the result.
`FlatMapIterator` ist a non templated class, but `from_map_iterator()` would be a templated member function.
Because of that, the required explicit template instantiation can not be done via the `EXPLICITLY_INSTANTIATE_DATA_TYPES` macro (because it only works with classes) but needs to be done the following way:
```
template FlatMapIterator FlatMapIterator::create_iterator<int32_t>(const tsl::sparse_map<int, std::vector<RowID>>::const_iterator& it);
template FlatMapIterator FlatMapIterator::create_iterator<int64_t>(const tsl::sparse_map<long, std::vector<RowID>>::const_iterator& it);
template FlatMapIterator FlatMapIterator::create_iterator<float>(const tsl::sparse_map<float, std::vector<RowID>>::const_iterator& it);
template FlatMapIterator FlatMapIterator::create_iterator<double>(const tsl::sparse_map<double, std::vector<RowID>>::const_iterator& it);
template FlatMapIterator FlatMapIterator::create_iterator<pmr_string>(const tsl::sparse_map<pmr_string, std::vector<RowID>>::const_iterator& it);
```
Also we would need to pass the DataType as a template to every function call:
`return Iterator::create_iterator<DataType>(_positions.cend());`.

I also think it would be nicer to have the Impl capsuled inside the `FlatMapIterator`, but I think the disadvantages that come with it are not worth it.

Reply:
If you worry about the (ugly) instantiation, we do the same with a macro and boost in base_test.cpp:
https://github.com/hyrise/hyrise/blob/41512c379b874f51b3d202608854046959cb6a4a/src/test/base_test.cpp#L41-L45

So would roughly be:
```c++
#define INSTANTIATE_CREATE_ITERATOR(r, data, type)                         \ 
  template FlatMapIterator FlatMapIterator::create_iterator<type>(         \
      const tsl::sparse_map<type, std::vector<RowID>>::const_iterator& it); 
  
BOOST_PP_SEQ_FOR_EACH(INSTANTIATE_CREATE_ITERATOR, _, DATA_TYPES) 
```

Think that's somehow okayish and a template parameter for a function call (inside a method of a templated class) is not a reasonable downside IMO.

Reply:
Another (I think over the top in this case) solution might be found in segment_accessor.hpp/cpp
https://github.com/hyrise/hyrise/blob/41512c379b874f51b3d202608854046959cb6a4a/src/lib/storage/segment_accessor.hpp#L16-L35

Reply:
This definitely looks better than my approach, but I am afraid it is not really good looking either.
Is it really so beneficial to capsule the Impl inside the Iterator when the PartialHashIndex and its iterator are closely interwoven anyway?
To me, both solutions are equally nice / unnice.
What do the others think about this? @dey4ss @Bouncner

Reply:
IMHO no other class should care about any other class using the pImpl pattern and actually initialize this impl. You can still outvote me :D

Reply:
I'd prefer not to add just another data type-specific macro (i.e., INSTANTIATE_CREATE_ITERATOR), and the `CreateSegmentAccessor` is quite verbose.

Since there is a need for `EXPLICITLY_INSTANTIATE_DATA_TYPES` for functions in other code locations, we could also add another macro for that in `all_type_variant.hpp`.

Reply:
I looked into this a couple of years ago for that occurence in base_test.cpp. Remember that it was not easy since the data type can appear multiple times in function signatures (return type, [multiple] parameters, ...) and it can quickly become uglier than before. Not really spent much afford back then and am not totally into macro writing though.

Reply:
I just implemented the solution segment_accessor used.
How do you like this way?

Reply:
Hm, no strong opinion if this is better than the macro or explicit instantiation. However, I think the `FlatMapIterator` constructor w/ impl as parameter can be protected now.

Reply:
I also don't have a strong opinion, but I slightly prefer the current solution and manual instantiation over the `INSTANTIATE_CREATE_ITERATOR` macro.
@tjjordan since you prefer the manual instantiation, feel free to change it accordingly.

Reply:
I just pushed a solution with manual instantiation.
I also don't have a strong opinion on this matter, but I think this way it is a little bit better understandable as with an extra class.